p2p/swarm.go

```diff
@@ -877,3 +871,77 @@ func (s *swarm) hasOutgoingPeer(peer p2pcrypto.PublicKey) bool {
 	s.outpeersMutex.RUnlock()
 	return ok
 }
+
+var listeningIp = inet.ParseIP("0.0.0.0")
+
+func (s *swarm) getListeners(
+	getTcpListener func(tcpAddr *inet.TCPAddr) (inet.Listener, error),
+	getUdpListener func(udpAddr *inet.UDPAddr) (net.UDPListener, error),
+	discoverUpnpGateway func() (nat_traversal.UpnpGateway, error),
+) (inet.Listener, net.UDPListener, error) {
+
+	port := s.config.TCPPort
+	randomPort := port == 0
+	var gateway nat_traversal.UpnpGateway
+	if s.config.AcquirePort {
+		var err error
+		gateway, err = discoverUpnpGateway()
+		if err != nil {
+			s.logger.With().Warning("could not discover UPnP gateway", log.Err(err))
+		}
+	}
+
+	for {
+		tcpAddr := &inet.TCPAddr{IP: listeningIp, Port: port}
+		tcpListener, err := getTcpListener(tcpAddr)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to acquire requested tcp port: %v", err)
+		}
+
+		addr := tcpListener.Addr()
+		port = addr.(*inet.TCPAddr).Port
+		udpAddr := &inet.UDPAddr{IP: listeningIp, Port: port}
+		udpListener, err := getUdpListener(udpAddr)
+		if err != nil {
+			if err := tcpListener.Close(); err != nil {
+				s.logger.With().Error("error closing tcp listener", log.Err(err))
+			}
+			if randomPort {
+				port = 0
+				continue
+			}
+			return nil, nil, fmt.Errorf("failed to acquire requested udp port: %v", err)
+		}
+
+		if gateway != nil {
+			err := nat_traversal.AcquirePortFromGateway(gateway, uint16(port))
+			if err != nil {
+				if err := tcpListener.Close(); err != nil {
+					s.logger.With().Error("error closing tcp listener", log.Err(err))
+				}
+				if err := udpListener.Close(); err != nil {
+					s.logger.With().Error("error closing udp listener", log.Err(err))
+				}
+				if randomPort {
+					port = 0
+					continue
+				}
+				return nil, nil, fmt.Errorf("failed to acquire requested port using UPnP: %v", err)
+			}
+		}
+
+		return tcpListener, udpListener, nil
+	}
+}
+
+func getUdpListener(udpAddr *inet.UDPAddr) (net.UDPListener, error) {
+	return inet.ListenUDP("udp", udpAddr)
+}
+
+func getTcpListener(tcpAddr *inet.TCPAddr) (inet.Listener, error) {
+	return inet.Listen("tcp", tcpAddr.String())
+}
+
+func discoverUpnpGateway() (igd nat_traversal.UpnpGateway, err error) {
```

Comment:
```suggestion
func discoverUPnPGateway() (igd nat_traversal.UpnpGateway, err error) {
```

lets try to keep acronyms with caps or the way they are written.